packages/apollo/src/commands/service/codegen.ts

```diff
@@ -0,0 +1,129 @@
+> = {
+  typescript: {
+    humanReadable: "TypeScript",
+    extension: "ts"
+  },
+  ts: {
+    humanReadable: "TypeScript",
+    extension: "ts"
+  }
+};
+
+const fileTypesMapping = {
+  ".ts": "javascript",
+  ".tsx": "javascript",
+  ".js": "javascript",
+  ".jsx": "javascript",
+  ".gql": "graphql",
+  ".graphql": "graphql"
+};
+
+export default class ServiceCodegen extends Command {
+  static description =
+    "Generate resolver typings from a local schema file. This can be either a schema embedded in `gql` tags in a JavaScript or TypeScript file, or a `.graphql` file containing the service's SDL.";
+  static flags = {
+    watch: flags.boolean({
+      char: "w",
+      description: "Watch the specified file for changes"
+    }),
+    target: flags.string({
+      char: "t",
+      description:
+        "Language to build typings for (currently only typescript is supported)",
+      default: "typescript"
+    }),
+    output: flags.string({
+      char: "o",
+      description: "Name of the file to write generated typings to",
+      default: "resolver-types"
+    })
+  };
+
+  static examples = [
+    "apollo service:codegen schema.ts -o schema.d.ts",
+    "apollo service:codegen schema.gql -o schema.d.ts"
+  ];
+
+  static args = [{ name: "input", required: true }];
+
+  async run() {
+    const { flags, args } = this.parse(ServiceCodegen);
+
+    const names = namesMapping[flags.target!];
+    if (!names) {
+      throw new Error(
+        `Unsupported target "${
+          flags.target
+        }. Supported values are: "${Object.keys(namesMapping).join(", ")}"`
+      );
+    }
+    const target = flags.target as Language;
+
+    const input: string = args.input;
+
+    const output =
+      (flags.output as string).indexOf(".") > -1
+        ? flags.output!
+        : `${flags.output}.${names.extension}`;
+
+    this.log(
+      `Generating ${names.humanReadable} typings: ${input} -> ${output}`
+    );
+
+    const inputPath = join(process.cwd(), input);
+    await this.executeCodegen(inputPath, target, output);
+
+    if (flags.watch) {
+      this.log(`Watching for changes...`);
+      watch(inputPath, async () => {
+        this.log(`Change detected. Regenerating typings.`);
+        try {
+          await this.executeCodegen(inputPath, target, output);
+        } catch (e) {
+          this.warn("Unable to run codegen: " + e.message);
+        }
+      });
+    }
+  }
+
+  private async executeCodegen(path: string, target: Language, output: string) {
+    const inputText = await new Promise<string>(resolve =>
+      readFile(path, (err, data) => {
+        if (err) throw new Error(JSON.stringify(err));
+        resolve(data.toString());
+      })
+    );
+    const inputFileType = extname(path);
+    const getSDL = () => {
+      const type = fileTypesMapping[inputFileType];
+      switch (type) {
+        case "javascript":
+          const getInnerSDL = /(?:(?:gql|graphql)`)((?:[^`\\]|\\`|\\n|\\r|\\\\)*)`/;
```

Comment:
I believe this is the approach our codegen libraries use now, but it would be great to start using ASTs for this. If you're not opposed to making this change, I'd be happy to help - though I haven't done it myself (but have wanted to!)

@benjamn will know the most. AST via `recast` and `visit()` from `ast-types`?

Reply:
The following should work, though I strongly recommend typing it in yourself, so you can see how much help TypeScript is providing:
```js
import { parse } from "recast/parsers/babel";
import { visit, types } from "recast";

const { Identifier } = types.namedTypes;
const inputText = "gql`query { whatever }`";
const operations: string[] = [];

visit(parse(inputText), {
  visitTaggedTemplateExpression(path) {
    this.traverse(path);
    const expr = path.node;
    if (
      Identifier.check(expr.tag) &&
      expr.tag.name === "gql"
    ) {
      expr.quasi.quasis.forEach(q => {
        operations.push(q.value.raw);
      });
    }
  },
});
```
I don't know how anyone writes AST transformations without auto-generated AST type information.

Reply:
Got it, thanks!